### PR TITLE
[mlir][linalg] Switch to use OpOperand* in ControlPropagationFn.

### PR DESCRIPTION
**Pull request by hanhanW**
It's not easy to determine whether we want to propagate pack/unpack ops because we don't know the (producer, consumer) information. The revisions switch it to `OpOperand*`, so the control function can capture the (producer, consumer) pair. E.g.,

```
Operation *producer = opOperand->get().getDefiningOp();
Operation *consumer = opOperand->getOwner();
```

### CODE REVIEW

**Review by MaheshRavishankar — COMMENTED** [review]
I have found it convenient to just use `OpOperand*` . That way the caller doesn't have to pass producer, consumer explicitly

**Comment by hanhanW** [comment]
> I have found it convenient to just use `OpOperand*` . That way the caller doesn't have to pass producer, consumer explicitly

I don't follow how it works with a single `OpOperand*`. Can you elaborate a little more? In the data layout propagation case, there could be `(pack, reshape)`, `(reshape, pack)`, `(generic, pack)`, `(pack, generic)`, etc. I don't follow how to distinguish `(pack, reshape)` and `(reshape, pack)` cases in the approach.

**Comment by MaheshRavishankar** [comment]
> > I have found it convenient to just use `OpOperand*` . That way the caller doesn't have to pass producer, consumer explicitly
> 
> I don't follow how it works with a single `OpOperand*`. Can you elaborate a little more? In the data layout propagation case, there could be `(pack, reshape)`, `(reshape, pack)`, `(generic, pack)`, `(pack, generic)`, etc. I don't follow how to distinguish `(pack, reshape)` and `(reshape, pack)` cases in the approach.

From the `OpOperand *` you can do this
```
Operation *producer = opOperand->get().getDefiningOp();
Operation *consumer = opOperand->getOwner();
```

For the `(pack, reshape)` case, the producer is a `pack` and consumer is a `reshape`. For the `(reshape, pack)` case the producer is a `reshape` and consumer is `pack`.

**Comment by Max191** [comment]
+1 to using `OpOperand *`. It is more consistent with other propagation patterns and less to be passing around.

**Review by rengolin — COMMENTED** [review]
This looks more robust, yeah.

@adam-smnk, I believe this won't change a single thing for us, right?

File: mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp [hunk 1]
```diff
@@ -879,10 +879,13 @@ class BubbleUpPackOpThroughReshapeOp final
...
 ///     : tensor<?x32x32x8x8xf32> -> tensor<?x256x256xf32>
-static LogicalResult
-pushDownUnPackOpThroughExpandShape(tensor::UnPackOp unPackOp,
-                                   tensor::ExpandShapeOp expandOp,
-                                   PatternRewriter &rewriter) {
+static LogicalResult pushDownUnPackOpThroughExpandShape(
+    tensor::UnPackOp unPackOp, tensor::ExpandShapeOp expandOp,
+    PatternRewriter &rewriter, ControlPropagationFn controlFn) {
+  // User controlled propagation function.
+  if (!controlFn(expandOp.getSrcMutable()))
```
Comment (rengolin) [inline mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:886]:
```suggestion
  if (!controlFn(&expandOp.getSrcMutable()))
```

**Comment by adam-smnk** [comment]
> I believe this won't change a single thing for us, right?

Functionally no changes, just a small update on how we call the pattern.
Overall +1 from me as it provides more control.

File: mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp (hunk 1 above)
Comment (rengolin) [reply mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:886]: This is what the test breaks about. Needs to be fixed before merging.

Comment (hanhanW) [reply mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp]: It seems like I forgot to add the change to the commit before I push the changes... Thanks for helping on this.

**Review by rengolin — APPROVED** [review]
LGTM, thanks!

**Pull request merged**